File: java/patterns/src/main/java/org/prototypepatterns/creational/abstractfactory/AbstractFactoryPattern.java
package org.prototypepatterns.creational.abstractfactory;

import java.util.List;

public final class AbstractFactoryPattern {
    private AbstractFactoryPattern() {
    }

    interface Button {
        String paint();
    }

    interface Checkbox {
        String paint();
    }

    interface ThemeFactory {
        Button createButton();

        Checkbox createCheckbox();
    }

    static final class DarkButton implements Button {
        @Override
        public String paint() {
            return "dark-button";
        }
    }

    static final class DarkCheckbox implements Checkbox {
        @Override
        public String paint() {
            return "dark-checkbox";
        }
    }

    static final class DarkThemeFactory implements ThemeFactory {
        @Override
        public Button createButton() {
            return new DarkButton();
        }

        @Override
        public Checkbox createCheckbox() {
            return new DarkCheckbox();
        }
    }

    public static List<String> demo() {
        ThemeFactory factory = new DarkThemeFactory();
        return List.of(factory.createButton().paint(), factory.createCheckbox().paint());
    }
}

File: csharp/patterns/src/PrototypePatterns/Creational/AbstractFactory/AbstractFactoryPattern.cs
namespace PrototypePatterns.Creational.AbstractFactory;

public interface IThemeButton
{
    string Paint();
}

public interface IThemeCheckbox
{
    string Paint();
}

public interface IThemeFactory
{
    IThemeButton CreateButton();
    IThemeCheckbox CreateCheckbox();
}

public sealed class DarkButton : IThemeButton
{
    public string Paint() => "dark-button";
}

public sealed class DarkCheckbox : IThemeCheckbox
{
    public string Paint() => "dark-checkbox";
}

public sealed class DarkThemeFactory : IThemeFactory
{
    public IThemeButton CreateButton() => new DarkButton();

    public IThemeCheckbox CreateCheckbox() => new DarkCheckbox();
}

public static class AbstractFactoryPattern
{
    public static IReadOnlyList<string> Demo()
    {
        IThemeFactory factory = new DarkThemeFactory();
        return [factory.CreateButton().Paint(), factory.CreateCheckbox().Paint()];
    }
}
